benchmarks/python/rope_ops.py

```diff
@@ -0,0 +1,1115 @@
+        cos = torch.randn(
+            cfg.batch_size,
+            cfg.seq_len,
+            head_dim,
+            device="cuda",
+            dtype=torch.bfloat16,
+            requires_grad=True,
+        )
+        sin = torch.randn(
+            cfg.batch_size,
+            cfg.seq_len,
+            head_dim,
+            device="cuda",
+            dtype=torch.bfloat16,
+            requires_grad=True,
+        )
+        return q, k, v, cos, sin
+
+    def grads():
+        grad = torch.randn(
+            cfg.batch_size,
+            cfg.num_attention_heads,
+            cfg.seq_len,
+            head_dim,
+            device="cuda",
+            dtype=torch.bfloat16,
+            requires_grad=False,
+        )
+        return grad
+
+    # Manual IOBytes computes the total bandwidth for thunder backward trace.
+    def iobytes():
+        n_elements = 0
+        # adding size of query_states.grad + key_states.grad + value_states.grad
+        n_elements += (
+            3 * cfg.batch_size * cfg.num_attention_heads * cfg.seq_len * head_dim
+        )
+        # adding size of query_states + key_states
+        n_elements += (
+            2 * cfg.batch_size * cfg.num_attention_heads * cfg.seq_len * head_dim
+        )
+        # adding size of cos, sin
+        n_elements += 2 * cfg.batch_size * cfg.seq_len * head_dim
+        # adding size of q.grad
+        n_elements += cfg.batch_size * cfg.seq_len * cfg.num_attention_heads * head_dim
+        # adding size of k.grad, v.grad
+        n_elements += (
+            2 * cfg.batch_size * cfg.seq_len * cfg.num_key_value_heads * head_dim
+        )
+        # adding size of cos.grad, sin.grad
+        n_elements += 2 * cfg.batch_size * cfg.seq_len * head_dim
+        # scale by dtype size
+        return n_elements * torch.bfloat16.itemsize
+
+    return Qwen2Rope(cfg).cuda().bfloat16(), inputs, grads, iobytes
+
+
+def hf_phi3_rope():
+    import json
+    from transformers.models.phi3 import Phi3Config
+
+    phi35_cfg_str = r"""{
+      "_name_or_path": "microsoft/Phi-3.5-mini-instruct",
+      "architectures": [
+        "Phi3ForCausalLM"
+      ],
+      "attention_bias": false,
+      "attention_dropout": 0.0,
+      "auto_map": {
+        "AutoConfig": "microsoft/Phi-3.5-mini-instruct--configuration_phi3.Phi3Config",
+        "AutoModelForCausalLM": "microsoft/Phi-3.5-mini-instruct--modeling_phi3.Phi3ForCausalLM"
+      },
+      "bos_token_id": 1,
+      "embd_pdrop": 0.0,
+      "eos_token_id": 32000,
+      "hidden_act": "silu",
+      "hidden_size": 3072,
+      "initializer_range": 0.02,
+      "intermediate_size": 8192,
+      "max_position_embeddings": 131072,
+      "model_type": "phi3",
+      "num_attention_heads": 32,
+      "num_hidden_layers": 32,
+      "num_key_value_heads": 32,
+      "original_max_position_embeddings": 4096,
+      "pad_token_id": 32000,
+      "resid_pdrop": 0.0,
+      "rms_norm_eps": 1e-05,
+      "rope_scaling": {
+        "long_factor": [
```

Comment:
Where do these numbers come from? Can you add a reference if there is one?

Reply:
Those were just dumped config from the model that we pulled.

```
from transformers import AutoModelForCausalLM
model = AutoModelForCausalLM.from_pretrained("microsoft/Phi-3.5-mini-instruct")
print(model.config)
```

So we don't have to pull the real model. Does this sound right to you @kevinstephano ?